shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java

```diff
@@ -898,6 +906,28 @@ void onUserLeaveHint() {
     }
   }
 
+  /**
+   * Invoke this from {@code Activity#onWindowFocusChanged()}.
+   *
+   * <p>A {@code Fragment} host must have its containing {@code Activity} forward this call so that
+   * the {@code Fragment} can then invoke this method.
+   */
+  void onWindowFocusChanged(boolean hasFocus) {
+    ensureAlive();
+    Log.v(TAG, "Received onWindowFocusChanged: " + (hasFocus ? "true" : "false"));
+    if (host.shouldDispatchAppLifecycleState() && flutterEngine != null) {
+      // This may notify in a redundant fashion with onResume and onPause. The
+      // lifecycle channel will deduplicate, but since the Android documentation
+      // says that we shouldn't rely on any particular order between lifecycle
+      // events and onWindowFocusChanged, we notify in both cases.
+      if (hasFocus) {
+        flutterEngine.getLifecycleChannel().appIsResumed();
+      } else {
+        flutterEngine.getLifecycleChannel().appIsInactive();
```

Comment:
I think this is really the core of the review. Do we want to overload inactive to mean either paused or no focus? 

My gut based on the bug is yes but I am worried about breaking existing apps assumptions. 

@gnprice Open to your thoughts here.

Reply:
That reasoning seems right to me.

I think it'd be good to amend that "On Android," paragraph in the [AppLifecycleState.inactive] doc a bit with this change, to make it better align with the terminology in the Android docs. I don't know what "the foreground inactive state" means in an Android context; but if it said "is paused, or is resumed but loses window focus", then that'd make it clear how to map it to the Android lifecycle methods.

Perhaps also add on [AppLifecycleState.resumed] a note that that state no longer agrees with the Android SDK's "resumed" concept, something like

> On Android, this state corresponds to an app or the Flutter host view having focus while in an Android "resumed" state.

That wording is perhaps still confusing and can probably be improved. But I think some confusion is inherent once we choose to have the state we call "resumed" not be the same on Android as the state Android calls "resumed", so we should try to be explicit about it.

Reply:
Yes, it's unfortunate that we have all these overloaded terms that mean different things on different platforms. For instance, that on Android we go into the `paused` state in `onStop`, and into the `inactive` state in `onPause` is particularly confusing. I'm not sure there's a cure for that, other than breaking all the apps (which we've elected not to do).

(More examples in this [design doc](https://docs.google.com/document/d/1FL5Fgk7EzMNOpJL0GZJt-lfWtGt2igHuC2ZznItS2l8/edit?resourcekey=0-IZNfpSVPkMbigQLLXoJcfw#bookmark=id.dtrwzax9ec0t).)

Added more explicit comments to the enum docs. Also, I have another doc update in this [draft PR that adds more app lifecycle support](https://github.com/flutter/engine/pull/40542/files#diff-57d6953e215d0e5dd7260ee60b665c812aa730566ef0b30f7ff1e3d661143585).

Reply:
Great, those added comments look good to me.

> I'm not sure there's a cure for that, other than breaking all the apps (which we've elected not to do).

Yeah. Given the differences between platforms, even if we found the best possible set of names unconstrained by compatibility with existing apps and used those, it'd probably still be fairly confusing. So I think the best we can do is just try to be explicit in the docs about what they do mean in each platform's terms.